### PR TITLE
pallet-xcm docs

### LINKED ISSUE #17: XCM Ingest + Local Execution Research
Closed by #76 #77

### PR DESCRIPTION
**Pull request by peektism**
Closes #17

### CODE REVIEW

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+* Pallet-specific ID types:
+** `RemoteLockConsumerIdentifier` -- The ID type for local consumers of remote locks. The type must implement `Copy`.
+
+* Pallet-specific handlers:
+** `OnChargeTransaction` -- Handler for withdrawing, refunding and depositing the transaction fee. Type must implement the trait  `OnChargeTransaction<Self>`.
+** `Currency` -- Lockable currency used in the pallet. Type must implement the trait `LockableCurrency`.
+** `Weigher` -- Measures weight consumed by XCM local execution. Type must implement the trait `WeightBounds`.
+** XcmExecutor -- Implements XCM messages. Type must implement the trait `ExecuteXcm`.
+** XcmRouter --  The type used to dispatch an XCM to its destination. Type must implement the trait `SendXcm`.
```
Comment (KitHat) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:39]: Minor formatting problems, code block is missing

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+----
+Execute an XCM message from a local, signed, origin.
+
+The origin must be `ExecuteXcmOrigin` for this call.
+
+NOTE: A successful return to this does NOT imply that the `msg` was executed successfully to completion; only that SOME of it was executed.
+
+**Params:**
+
+- `message: Box<VersionedXcm<<T as Config>::RuntimeCall>>` — versioned XCM to be executed
```
Comment (KitHat) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:118]: There is some weird glitch in compiled version, it is rendered as a link.
Could you please try to fix it?

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 3]
```diff
@@ -0,0 +1,340 @@
+:source-highlighter: highlight.js
+:highlightjs-languages: rust
+:github-icon: pass:[<svg class="icon"><use href="#github-icon"/></svg>]
+
+= pallet_xcm
+
+Branch/Release: `release-polkadot-v{1.3.0}`
```
Comment (KitHat) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:7]: Could you please remove curly braces?

**Review by KitHat — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+== Purpose
+
+`pallet-xcm` is responsible for filtering, routing, and executing incoming XCM.
+
+== Glossary
+
+** `Barrier` -- whether to execute the XCM call at all
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:15]: this explanation is a bit confusing imo. The reason for me is: `whether to execute the XCM call at all` is a boolean, yet `barrier` is probably not a boolean. 

If `barrier` is what I think, maybe rewording this as `set of criteria on whether to execute the XCM calls` would be better.

I may have misinterpreted the whole thing, feel free to correct me

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+** `Fungible Asset` -- any unit is worth the same
+** `NonFungible Asset` -- each unit is worth a different amount and/or is unique in some way
+** `Junction` -- single item in a path to describe a relative location (think of `dir` or `dir/$file_path` as a junction)
+** `Junctions` -- multiple `Junction`s, formed only through MultiLocation construction
+** `MultiAsset` -- Either an amount of a single fungible asset, or a single well-identified non-fungible asset.
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:20]: what does `well-identified` mean in this context? It might be confusing for some

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+** `MultiLocation` -- path described by junctions leading to the location
+** `MAX_ASSETS_FOR_TRANSFER` -- The maximum number of distinct assets allowed to be transferred in a single helper extrinsic. Set to 2 in this code.
+
+== Config link:https://github.com/paritytech/polkadot-sdk/blob/release-polkadot-v1.3.0/substrate/frame/pallet-xcm/src/lib.rs#L192[{github-icon},role=heading-link]
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:24]: link is broken

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+** `SendXcmOrigin` -- Required origin for sending XCM messages. If successful, it resolves to `MultiLocation`. Type must implement trait `EnsureOrigin`.
+
+* Pallet-specific ID types:
+** `RemoteLockConsumerIdentifier` -- The ID type for local consumers of remote locks. The type must implement `Copy`.
+
+* Pallet-specific handlers:
+** `OnChargeTransaction` -- Handler for withdrawing, refunding and depositing the transaction fee. Type must implement the trait  `OnChargeTransaction<Self>`.
+** `Currency` -- Lockable currency used in the pallet. Type must implement the trait `LockableCurrency`.
+** `Weigher` -- Measures weight consumed by XCM local execution. Type must implement the trait `WeightBounds`.
+** XcmExecutor -- Implements XCM messages. Type must implement the trait `ExecuteXcm`.
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:38]: > Implements XCM messages

is a bit misleading imo. I prefer the original explanation here instead: `Something to execute an XCM message.`, this was less confusing at least for me.

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+** XcmRouter --  The type used to dispatch an XCM to its destination. Type must implement the trait `SendXcm`.
+
+* Pallet-specific filters:
+** `XcmExecuteFilter` -- XCM filter for which messages to be executed using `XcmExecutor` must pass. Type must implement trait `Contains<(MultiLocation, Xcm<<Self as Config>::RuntimeCall>)>`.
+** `XcmReserveTransferFilter` -- XCM filter for which messages to be reserve-transferred using the dedicated extrinsic must pass.Type must implement the trait `Contains<(MultiLocation, Vec<MultiAsset>)>`.
+** `XcmTeleportFilter` -- XCM filter for which messages to be teleported using the dedicated extrinsic must pass. Type must implement the trait `Contains<(MultiLocation, Vec<MultiAsset>)>`.
+** `TrustedLockers` -- The assets which we consider a given origin is trusted if they claim to have placed a lock. Type must implement `ContainsPair`.
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:45]: I know this is the original explanation, but don't you think it is hard to understand and cryptic? 

I think we can improve this explanation

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+* Pallet-specific converters:
+** `CurrencyMatcher` -- The `MultiAsset` matcher for `Currency`. Type must implement the trait `MatchesFungible`.
+** `SovereignAccountOf` -- How to get an `AccountId` value from a `MultiLocation`, useful for handling asset locks. Type must implement `ConvertLocation`.
+
+* Pallet-specific constants:
+** `VERSION_DISCOVERY_QUEUE_SIZE` -- `u32` measures the size of the version discovery queue. Typically to set to `100`.
```
Comment (ozgunozerk) [inline docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:52]: does it measure, or set? 

asking since I don't know, but the naming felt like it is setting (i know that sometimes parity is not doing great with naming, just double-checking)

File: docs/modules/ROOT/pages/pallets/pallet_xcm.adoc [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+** `Junctions` -- multiple `Junction`s, formed only through MultiLocation construction
+** `MultiAsset` -- Either an amount of a single fungible asset, or a single well-identified non-fungible asset.
+** `MultiLocation` -- path described by junctions leading to the location
+** `MAX_ASSETS_FOR_TRANSFER` -- The maximum number of distinct assets allowed to be transferred in a single helper extrinsic. Set to 2 in this code.
+
+== Config link:https://github.com/paritytech/polkadot-sdk/blob/release-polkadot-v1.3.0/substrate/frame/pallet-xcm/src/lib.rs#L192[{github-icon},role=heading-link]
+
+* Pallet-specific origins:
+** `AdminOrigin` -- The origin that is allowed to call privileged operations on the XCM pallet. Type must implement trait `EnsureOrigin`.
+** `ExecuteXcmOrigin` -- Required origin for executing XCM messages, including the teleport functionality. If successful, it returns a `MultiLocation` which exists as an interior location within this chain's XCM context. Type must implement trait `EnsureOrigin`.
```
[the remark this answers is not in the document]
Comment (ozgunozerk) [reply docs/modules/ROOT/pages/pallets/pallet_xcm.adoc:28]: I learned what it is after I read the dispatchable about the teleport. 
But this in the very end of the documentation. 

It's up to you to decide whether to provide a small intro to `teleport` concept in the beginning.

Feel free to resolve this as you like 👍

**Review by ozgunozerk — CHANGES_REQUESTED** [review]
LGTM overall, wonderful work!

few more clarifications would make it superb imo

**Review by ozgunozerk — APPROVED** [review]
💯

**Review by KitHat — APPROVED** [review]
